scalalib/src/TestModule.scala

```diff
@@ -0,0 +1,215 @@
+   * */
+  def testUseArgsFile: T[Boolean] = T { scala.util.Properties.isWin }
+
+  protected def testTask(
+      args: Task[Seq[String]]): Task[(String, Seq[TestRunner.Result])] =
+    T.task {
+      val outputPath = T.dest / "out.json"
+      val useArgsFile = testUseArgsFile()
+
+      val (jvmArgs, props: Map[String, String]) = if (useArgsFile) {
+        val (props, jvmArgs) = forkArgs().partition(_.startsWith("-D"))
+        val sysProps =
+          props
+            .map(_.drop(2).split("[=]", 2))
+            .map {
+              case Array(k, v) => k -> v
+              case Array(k)    => k -> ""
+            }
+            .toMap
+
+        jvmArgs -> sysProps
+      } else {
+        forkArgs() -> Map()
+      }
+
+      val testArgs = TestRunner.TestArgs(
+        framework = testFramework(),
+        classpath = runClasspath().map(_.path.toString()),
+        arguments = args(),
+        sysProps = props,
+        outputPath = outputPath.toString(),
+        colored = T.log.colored,
+        testCp = compile().classes.path.toString(),
+        homeStr = T.home.toString()
+      )
+
+      val mainArgs = if (useArgsFile) {
+        val argsFile = T.dest / "testargs"
+        Seq(testArgs.writeArgsFile(argsFile))
+      } else {
+        testArgs.toArgsSeq
+      }
+
+      Jvm.runSubprocess(
+        mainClass = "mill.scalalib.TestRunner",
+        classPath = zincWorker.scalalibClasspath().map(_.path),
+        jvmArgs = jvmArgs,
+        envArgs = forkEnv(),
+        mainArgs = mainArgs,
+        workingDir = forkWorkingDir(),
+        useCpPassingJar = useArgsFile
+      )
+
+      if (!os.exists(outputPath)) Result.Failure("Test execution failed.")
+      else
+        try {
+          val jsonOutput = ujson.read(outputPath.toIO)
+          val (doneMsg, results) =
+            upickle.default.read[(String, Seq[TestRunner.Result])](jsonOutput)
+          TestModule.handleResults(doneMsg, results)
+        } catch {
+          case e: Throwable =>
+            Result.Failure("Test reporting failed: " + e)
+        }
+    }
+
+  /**
+   * Discovers and runs the module's tests in-process in an isolated classloader,
+   * reporting the results to the console
+   */
+  def testLocal(args: String*) = T.command {
+    val outputPath = T.dest / "out.json"
+
+    val (doneMsg, results) = TestRunner.runTestFramework(
+      TestRunner.framework(testFramework()),
+      runClasspath().map(_.path),
+      Agg(compile().classes.path),
+      args,
+      T.testReporter
+    )
+
+    TestModule.handleResults(doneMsg, results)
+
+  }
+}
+
+object TestModule {
+
+  /**
+   * TestModule using TestNG Framework to run tests.
+   * You may want to provide the testng dependency explicitly to use another version.
+   */
+  trait TestNg extends TestModule {
+    override def ivyDeps: T[Agg[Dep]] = T {
+      super.ivyDeps() ++ Agg(
+        ivy"com.lihaoyi:mill-contrib-testng_2.13:${mill.BuildInfo.millVersion}")
```

Comment:
https://github.com/com-lihaoyi/mill/pull/1268/files#diff-aa5cde7401023fc2ba3efedb50e338ac1643922517017182878165d8b05af73cR24
Here you have `mill-contrib-testng` without `_2.13` suffix. Why?

Reply:
I think the one in caffeine tests is unused, as the test framework in use is JUnit. We need to review caffeine tests further, to make sure the setup is correct. I remember that I hat lots of issues with caffeine builds when I added integration tests for `mill-cobtrib-testng`.

Reply:
FTR, https://github.com/com-lihaoyi/mill/pull/566#issuecomment-470440705